server/src/test/java/org/elasticsearch/cluster/coordination/ReconfiguratorTests.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.Version;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterState.VotingConfiguration;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.set.Sets;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptySet;
+import static org.elasticsearch.cluster.coordination.Reconfigurator.CLUSTER_MASTER_NODES_FAILURE_TOLERANCE;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.sameInstance;
+
+public class ReconfiguratorTests extends ESTestCase {
+
+    public void testReconfigurationExamples() {
+        for (int masterNodesFailureTolerance = 0; masterNodesFailureTolerance <= 2; masterNodesFailureTolerance++) {
+            check(nodes("a"), conf("a"), masterNodesFailureTolerance, conf("a"));
+            check(nodes("a", "b"), conf("a"), masterNodesFailureTolerance, conf("a"));
+            check(nodes("a", "b", "c"), conf("a"), masterNodesFailureTolerance, conf("a", "b", "c"));
+            check(nodes("a", "b", "c"), conf("a", "b"), masterNodesFailureTolerance, conf("a", "b", "c"));
+            check(nodes("a", "b", "c"), conf("a", "b", "c"), masterNodesFailureTolerance, conf("a", "b", "c"));
+            check(nodes("a", "b", "c", "d"), conf("a", "b", "c"), masterNodesFailureTolerance, conf("a", "b", "c"));
+            check(nodes("a", "b", "c", "d", "e"), conf("a", "b", "c"), masterNodesFailureTolerance, conf("a", "b", "c", "d", "e"));
+            check(nodes("a", "b"), conf("a", "b", "e"), masterNodesFailureTolerance,
+                masterNodesFailureTolerance == 1 ? conf("a", "b", "e") : conf("a"));
```

Comment:
It's a bit weird to read these tests as we handle the `config.getNodeIds().size() < 2 * masterNodesFailureTolerance + 1)` case within `check`. It's something that you have to constantly keep in mind while reading these tests.
I would prefer to test that separately and only have `check`s that actually lead to the desired target configuration.
This will result in a bit more duplication, but I think that's fine.

Reply:
Ok, I reworked the tests to avoid the invalid cases better.